### PR TITLE
Add support for Arm64 Catalyst on ARM Macs

### PR DESCRIPTION
**Pull request by terhechte**
This is an iteration on https://github.com/rust-lang/rust/pull/63467 which was merged a while ago. In the aforementioned PR, I added support for the `X86_64-apple-ios-macabi` target triple, which is Catalyst, iOS apps running on macOS.

Very soon, Apple will launch ARM64 based Macs which will introduce `aarch64_apple_darwin.rs`, macOS apps using the Darwin ABI running on ARM. This PR adds support for Catalyst apps on ARM Macs: iOS apps compiled for the darwin ABI.

I don't have access to a Apple Developer Transition Kit (DTK), so I can't really test if the generated binaries work correctly. I'm vaguely hopeful that somebody with access to a DTK could give this a spin.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
cc @shepmaster -- perhaps you know of a good reviewer here due to Apple Silicon work

File: compiler/rustc_codegen_ssa/src/back/link.rs
Comment (shepmaster) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2104]: This doesn't have `x` like the similar target below. How can we double check this value?

File: compiler/rustc_target/src/spec/apple_sdk_base.rs
Comment (shepmaster) [inline compiler/rustc_target/src/spec/apple_sdk_base.rs:24]: I'm surprised by this (and the other instances) because the DTK (and presumably the real commercial releases) are `apple-a12` at a minimum.

File: compiler/rustc_target/src/spec/aarch64_apple_ios_macabi.rs
Comment (shepmaster) [inline compiler/rustc_target/src/spec/aarch64_apple_ios_macabi.rs:33]: It would be good to try and get a set of updated values for this, but I don't know how.

**Review by shepmaster — COMMENTED** [review]
> I'm vaguely hopeful that somebody with access to a DTK could give this a spin.

I have access. What steps are required?

File: compiler/rustc_target/src/spec/apple_sdk_base.rs
Comment (terhechte) [reply compiler/rustc_target/src/spec/apple_sdk_base.rs]: Thanks! I'd overlooked this. I've updated it accordingly to a12

File: compiler/rustc_target/src/spec/aarch64_apple_ios_macabi.rs
Comment (terhechte) [reply compiler/rustc_target/src/spec/aarch64_apple_ios_macabi.rs]: Thanks for pointing this out! I had to remove the `-target-abi darwinpcs` line in order for the libraries to be accepted by Xcode.

File: compiler/rustc_codegen_ssa/src/back/link.rs
Comment (terhechte) [reply compiler/rustc_codegen_ssa/src/back/link.rs]: You were right, it was the wrong value. I've updated it accordingly.

**Comment by terhechte** [comment]
> > I'm vaguely hopeful that somebody with access to a DTK could give this a spin.
> 
> I have access. What steps are required?

@shepmaster 

I've set up a demo project here https://github.com/terhechte/rust-catalyst-example
This includes a readme that explains how to build this against a local checkout of this PR. However, it also already contains pre-build binaries against this PR and an Xcode Project that integrates them. As Xcode doesn't complain (anymore) when building, I'm hopeful that the resulting binaries are valid.

In order to test this, you'd need to build the `XcodeIntegration.xcodeproj`, take the resulting app to a DTK and try to run it. It should print something on the console. 

Some Caveats:

1. You might need to code sign the binary. Adhoc signing should work: `codesign --force --deep -s - XcodeIntegration.app
2. You might want to build the libtest* binaries from hand and (just for security matters) not rely on my pre-compiled ones. The Readme.md in that folder explains the process (mostly just building this PR, Rustup'ing the toolchain and executing the `make_fat.sh` script). 

If you have any questions please reach out!

Thanks for looking into this!

**Comment by terhechte** [comment]
Good News! avbelow took the binaries that I build with the custom toolchain in this PR and ran them on his DTK. 

https://twitter.com/avbelow/status/1312746147623120897?s=21

So it seems this is working as intended :)

**Comment by below** [comment]
Can confirm. I have tested this on an Apple Silicon machine and it is working as expected

**Comment by Mark-Simulacrum** [comment]
@pietroalbini @joshtriplett -- do we have process in place for tier 3 target additions? I forget what we've done in the past and don't see any docs at a quick glance.

**Comment by pietroalbini** [comment]
@Mark-Simulacrum it needs approval from someone on the compiler team.

**Comment by Mark-Simulacrum** [comment]
Let's r? @nikomatsakis for T-compiler approval then.

File: src/doc/rustc/src/platform-support.md
Comment (shepmaster) [inline src/doc/rustc/src/platform-support.md:150]: Misapplied conflict resolution

Comment (terhechte) [reply src/doc/rustc/src/platform-support.md:149]: Sorry ☺️ Fixed it

Comment (shepmaster) [reply src/doc/rustc/src/platform-support.md:149]: It appears that your PR adds the Apple-Darwin target to this tier. It should not as it’s now in a higher tier.

Comment (terhechte) [reply src/doc/rustc/src/platform-support.md]: That was a remnant of the broken merge. Fixed that as well. Thanks!

**Review by nikomatsakis — APPROVED** [review]

**Comment by Dylan-DPC** [comment]
failed in [rollup CI](https://github.com/rust-lang/rust/pull/78032/checks?check_run_id=1267084601)

**Comment by terhechte** [comment]
@shepmaster What would be the next steps here to get this merged? There were merge conflicts that I resolved.

**Comment by shepmaster** [comment]
@bors r-

Before we merge this, let's clean up the commit history a bit. Too many merges of master for my tastes.

**Comment by nikomatsakis** [comment]
I'm going to r? @shepmaster for follow-up here, please feel free to ping me if further attention is required on my part

**Comment by terhechte** [comment]
@nikomatsakis @shepmaster I've cleaned up the history

**Pull request merged**
